### PR TITLE
Feature 1635 - customizing service name with prefix, flag --prefix (-x)

### PR DESCRIPTION
**Pull request by sosan**
#### What type of PR is this?

/kind feature 

#### What this PR does / why we need it:
The --prefix flag serves the purpose of allowing users to specify prefixes to be added to the names of services

#### Which issue(s) this PR fixes:

Feature #1635

#### Special notes for your reviewer:
I haven't had time to document the --prefix flag
documentation will be cooming soon in the following commits.

### CODE REVIEW

**Review by TessaIO — COMMENTED** [review]
Thanks for the patch @sosan. I left some comments below.

**Comment by cdrage** [comment]
@TessaIO shouldn't we have this as a label param instead? or do you think CLI parameter would be fine?

**Comment by TessaIO** [comment]
> @TessaIO shouldn't we have this as a label instead? or do you think CLI parameter would be fine?

That's actually true if we want to offer users flexibility to specify per-service prefixes that's the option to go (I mean labels here).

**Comment by cdrage** [comment]
> > @TessaIO shouldn't we have this as a label instead? or do you think CLI parameter would be fine?
> 
> That's actually true if we want to offer users flexibility to specify per-service prefixes that's the option to go (I mean labels here). wdyt @sosan?

Yeah just thinking about it again, I think I'm good for terminal / cli based prefixes being added, we could always customize the name individually via labels instead.

File: pkg/transformer/kubernetes/k8sutils_test.go [hunk 1]
```diff
@@ -738,3 +738,108 @@ func TestRemoveEmptyInterfaces(t *testing.T) {
...
+				komposeObject: &kobject.KomposeObject{
+					ServiceConfigs: map[string]kobject.ServiceConfig{"app": {Name: "WEB_API_BAD_NAME"}},
+				},
+			},
+			want: &kobject.KomposeObject{
+				ServiceConfigs: map[string]kobject.ServiceConfig{"app": {Name: "test-fail-web-api-bad-name"}},
+			},
+		},
+		{
+			name: "-Testing-Bad-",
```
[the remark this answers is not in the document]
Comment (sosan) [reply pkg/transformer/kubernetes/k8sutils_test.go]: done

File: pkg/transformer/kubernetes/k8sutils_test.go [hunk 2]
```diff
@@ -738,3 +738,108 @@ func TestRemoveEmptyInterfaces(t *testing.T) {
...
+				komposeObject: &kobject.KomposeObject{
+					ServiceConfigs: map[string]kobject.ServiceConfig{"app": {Name: "WEB_API_BAD_NAME"}},
+				},
+			},
+			want: &kobject.KomposeObject{
+				ServiceConfigs: map[string]kobject.ServiceConfig{"app": {Name: "test-fail-web-api-bad-name"}},
+			},
+		},
+		{
+			name: "-TEST_FAIL",
```
[the remark this answers is not in the document]
Comment (sosan) [reply pkg/transformer/kubernetes/k8sutils_test.go]: done

File: pkg/transformer/kubernetes/k8sutils.go [hunk 3]
```diff
@@ -985,3 +985,26 @@ func reformatSecretConfigUnderscoreWithDash(secretConfig types.ServiceSecretConf
...
+	return komposeObject
+}
+
+// It ensures consistency in the format of prefixes
+// prevents issues with leading or trailing dashes affecting resulting service names
+func cleanPrefix(prefix string) string {
+	prefix = FormatResourceName(prefix)
+	prefix = strings.TrimSuffix(prefix, "-")
+	prefix = strings.TrimPrefix(prefix, "-")
+	return prefix
```
[the remark this answers is not in the document]
Comment (sosan) [reply pkg/transformer/kubernetes/k8sutils.go]: done

File: pkg/transformer/kubernetes/k8sutils.go [hunk 4]
```diff
@@ -985,3 +985,26 @@ func reformatSecretConfigUnderscoreWithDash(secretConfig types.ServiceSecretConf
 
 	return newSecretConfig
 }
+
+// It takes a pointer to a KomposeObject and an array of prefixes
+// The first prefix from the array is used for appending
+// This function modifies the original KomposeObject directly
+func appendPrefixName(komposeObject *kobject.KomposeObject, prefixArray []string) *kobject.KomposeObject {
```
[the remark this answers is not in the document]
Comment (sosan) [reply pkg/transformer/kubernetes/k8sutils.go]: done

File: cmd/root.go [hunk 5]
```diff
@@ -111,4 +112,5 @@ func init() {
 	RootCmd.PersistentFlags().BoolVar(&GlobalErrorOnWarning, "error-on-warning", false, "Treat any warning as an error")
 	RootCmd.PersistentFlags().StringSliceVarP(&GlobalFiles, "file", "f", []string{}, "Specify an alternative compose file")
 	RootCmd.PersistentFlags().StringVar(&GlobalProvider, "provider", "kubernetes", "Specify a provider. Kubernetes or OpenShift.")
+	RootCmd.PersistentFlags().StringSliceVarP(&GlobalPrefixAdd, "prefix", "x", []string{}, "Add prefix in name services")
```
[the remark this answers is not in the document]
Comment (sosan) [reply cmd/root.go]: done

File: pkg/transformer/kubernetes/k8sutils_test.go [hunk 6]
```diff
@@ -738,3 +738,108 @@ func TestRemoveEmptyInterfaces(t *testing.T) {
...
+		komposeObject *kobject.KomposeObject
+		prefixArray   []string
+	}
+	tests := []struct {
+		name string
+		args args
+		want *kobject.KomposeObject
+	}{
+		{
+			name: "-TEST_FAIL-",
```
[the remark this answers is not in the document]
Comment (sosan) [reply pkg/transformer/kubernetes/k8sutils_test.go]: done

**Comment by sosan** [comment]
Updated cli

Sounds like a good idea to me. That way, we can fine-tune the names individually using labels

File: pkg/transformer/kubernetes/k8sutils.go [hunk 7]
```diff
@@ -985,3 +985,27 @@ func reformatSecretConfigUnderscoreWithDash(secretConfig types.ServiceSecretConf
...
+}
+
+// It ensures consistency in the format of prefixes
+// prevents issues with leading or trailing dashes affecting resulting service names
+func cleanPrefix(prefix string) string {
+	prefix = strings.Trim(FormatResourceName(prefix), "-")
+	//FormatResourceName(prefix)
+	// prefix = strings.TrimSuffix(prefix, "-")
+	// prefix = strings.TrimPrefix(prefix, "-")
+	return prefix
```
Comment (TessaIO) [inline pkg/transformer/kubernetes/k8sutils.go:1010]:
```suggestion
	return strings.Trim(FormatResourceName(prefix), "-")
```

File: pkg/app/app.go [hunk 8]
```diff
@@ -142,6 +142,7 @@ func ValidateFlags(args []string, cmd *cobra.Command, opt *kobject.ConvertOption
 		}
 		log.Fatal("Unknown Volume type: ", opt.Volumes, ", possible values are: ", strings.Join(validVolumesTypes, " "))
 	}
+
```
Comment (TessaIO) [inline pkg/app/app.go:145]:
```suggestion
```

**Review by TessaIO — COMMENTED** [review]
Small comments are left below.

File: pkg/app/app.go (hunk 8 above)
Comment (sosan) [reply pkg/app/app.go]: updated

**Review by TessaIO — APPROVED** [review]
LGTM! Leaving final approve to @cdrage

File: cmd/root.go [hunk 9]
```diff
@@ -111,4 +112,5 @@ func init() {
 	RootCmd.PersistentFlags().BoolVar(&GlobalErrorOnWarning, "error-on-warning", false, "Treat any warning as an error")
 	RootCmd.PersistentFlags().StringSliceVarP(&GlobalFiles, "file", "f", []string{}, "Specify an alternative compose file")
 	RootCmd.PersistentFlags().StringVar(&GlobalProvider, "provider", "kubernetes", "Specify a provider. Kubernetes or OpenShift.")
+	RootCmd.PersistentFlags().StringVarP(&GlobalPrefixAdd, "prefix", "x", "", "Add prefix in name services")
```
Comment (cdrage) [inline cmd/root.go:115]:
```suggestion
	RootCmd.PersistentFlags().StringVarP(&GlobalPrefixAdd, "prefix", "x", "", "Adds a custom prefix to service names")
```

**Review by cdrage — CHANGES_REQUESTED** [review]
One small change requested

**Review by cdrage — CHANGES_REQUESTED** [review]
If we are going to add prefix, we should also add suffix...

I do not believe we should have `-x` either. Prefix and suffix should honestly be fixed after generation. I am not opposed to having this merged in as there is an issue opened for it, but we should also add suffix too.

**Comment by sosan** [comment]
> we should also add suffix too.

ok, roger that

**Comment by TessaIO** [comment]
ping @cdrage

**Review by cdrage — CHANGES_REQUESTED** [review]

File: cmd/convert.go [hunk 10]
```diff
@@ -202,6 +203,7 @@ func init() {
 	convertCmd.Flags().BoolVar(&GenerateNetworkPolicies, "generate-network-policies", false, "Specify whether to generate network policies or not")
 
 	convertCmd.Flags().BoolVar(&WithKomposeAnnotation, "with-kompose-annotation", true, "Add kompose annotations to generated resource")
+	convertCmd.Flags().StringVarP(&GlobalPrefixAdd, "prefix", "x", "", "Adds a custom prefix to service names")
```
[the remark this answers is not in the document]
Comment (sosan) [reply cmd/convert.go:206]: Sure. Add a suffix with flag? Labels?

[the remark this answers is not in the document]
Comment (cdrage) [reply cmd/convert.go:206]: Just flag like how you did it with suffix

**Comment by cdrage** [comment]
Hey sorry @sosan for some reason I didn't realize that we reached a conclusion on https://github.com/kubernetes/kompose/issues/1635#issuecomment-2031874480 that we should actually close this PR, and I apologize the work you've put into this back and forth.

Going forward I do not think we should be adding too many suffix and prefix / name customizations unless it's "generic", if someone wants to edit the entire `-tcp` suffix, they could do it with a sed / ctrl +f / ctrl + replace after converting.

There's too much work and labels / parameters to be added to make something small like this happen.

**Pull request closed without merging**